mantis-server/mantis-server-worker/src/main/java/io/mantisrx/server/worker/TaskExecutor.java

```diff
@@ -0,0 +1,650 @@
+    private int resourceManagerCxnIdx;
+    private Throwable previousFailure;
+
+    public TaskExecutor(RpcService rpcService,
+                        WorkerConfiguration workerConfiguration,
+                        HighAvailabilityServices highAvailabilityServices, ClassLoaderHandle classLoaderHandle,
+                        Factory subscriptionStateHandlerFactory) {
+        super(rpcService, RpcServiceUtils.createRandomName("worker"));
+
+        // this is the task executor ID that will be used for the rest of the JVM process
+        this.taskExecutorID =
+            Optional.ofNullable(workerConfiguration.getTaskExecutorId())
+                .map(TaskExecutorID::of)
+                .orElseGet(TaskExecutorID::generate);
+        this.clusterID = ClusterID.of(workerConfiguration.getClusterId());
+        this.workerConfiguration = workerConfiguration;
+        this.highAvailabilityServices = highAvailabilityServices;
+        this.classLoaderHandle = classLoaderHandle;
+        this.subscriptionStateHandlerFactory = subscriptionStateHandlerFactory;
+        WorkerPorts workerPorts = new WorkerPorts(workerConfiguration.getMetricsPort(),
+            workerConfiguration.getDebugPort(), workerConfiguration.getConsolePort(),
+            workerConfiguration.getCustomPort(),
+            workerConfiguration.getSinkPort());
+        MachineDefinition machineDefinition = new MachineDefinition(
+            Hardware.getNumberCPUCores(),
+            Hardware.getSizeOfPhysicalMemory(),
+            Hardware.getSizeOfDisk(),
+            workerPorts.getNumberOfPorts());
+        this.taskExecutorRegistration =
+            new TaskExecutorRegistration(
+                taskExecutorID, clusterID, getAddress(), workerConfiguration.getExternalAddress(), workerPorts, machineDefinition);
+        this.ioExecutor =
+            Executors.newFixedThreadPool(
+                Hardware.getNumberCPUCores(),
+                new ExecutorThreadFactory("taskexecutor-io"));
+        this.resourceManagerCxnIdx = 0;
+    }
+
+    @Override
+    protected void onStart() throws Exception {
+        try {
+            startTaskExecutorServices();
+            startFuture.complete(null);
+        } catch (Throwable throwable) {
+            log.error("Fatal error occurred in starting TaskExecutor {}", getAddress(), throwable);
+            startFuture.completeExceptionally(throwable);
+            throw throwable;
+        }
+    }
+
+    private void startTaskExecutorServices() throws Exception {
+        validateRunsInMainThread();
+
+        masterMonitor = highAvailabilityServices.getMasterClientApi();
+        RxNetty.useMetricListenersFactory(new MantisNettyEventsListenerFactory());
+        establishNewResourceManagerCxnSync();
+        highAvailabilityServices
+            .connectWithResourceManager(clusterID)
+            .register(new ResourceManagerChangeListener());
+    }
+
+    public CompletableFuture<Void> awaitRunning() {
```

Comment:
Is this used? Cannot find any ref to it.

Reply:
This is used in the tests. In the tests, it's important to know when the task executor has started so that we can start issuing commands to it.